### PR TITLE
Add Decoding Verification for the Checkpoint Validation Pipeline

### PR DESCRIPTION
**Pull request by olufiyin19**
# Description

This PR introduces `decode_validator.py`, adding automated autoregressive decoding and generative inference validation to the automated post checkpoint conversion validation pipeline. This is the fourth phase in the stacked checkpoint validation series (building on PR #4462, PR #4726, and #4727).

While static shape checks and single-step forward-pass logit verification confirm that converted checkpoints (e.g. Gemma, Llama, Deepseek, Mistral, Qwen) align with expected network geometries and logit distributions, they do not validate multi-step autoregressive generation. Converted checkpoints can pass single-step forward checks yet still fail during text generation due to KV-cache allocation errors, RoPE/positional embedding misconfigurations, attention mask incompatibilities, or tokenizer decode incompatibilities.

This PR standardizes generative validation by utilizing MaxText's autoregressive decoding pipeline (`src/maxtext/inference/decode.py`), streaming subprocess stdout/stderr directly into application logs for remote observability, and publishing schema-compliant JSON reports (`report_{run_name}.json`) to Google Cloud Storage for Airflow fail-fast orchestration.

# Key Implementation Details

1. Autoregressive Decoding Orchestration (`validate_checkpoint`)
* Implemented an automated wrapper around `src/maxtext/inference/decode.py` that executes cleanly from the repository root directory (`repo_root`).
* Enforces strict validation of mandatory configuration overrides (`tokenizer_path` and `scan_layers`), raising immediate descriptive exceptions if required checkpoint metadata is omitted from Airflow or CLI overrides.

2. Real-Time Subprocess Logging & Traceability
* Captured full subprocess stdout and stderr streams during decode execution.
* Implemented explicit streaming of stdout (`=== Subprocess Stdout ===`) directly into `absl.logging` / MaxText logger so that generated prompt-response sequences (`Input ... -> ...`), token generation latency, and throughput metrics are visible in real time within Airflow and CI task logs.

3. Standardized JSON Schema Reporting & GCS Integration
* Formats decoding outcomes, generated text samples, execution status (`SUCCESS` vs `FAILED`), and error diagnostics into a structured JSON report (`report_{run_name}.json`).
* Automatically saves reports locally (`./reports/report_{run_name}.json`) and uploads them to Google Cloud Storage (`report_gcs_dir`), enabling Airflow fail-fast pipelines and the automated Fixer Agent Sidecar to inspect decoding fidelity and trigger self-healing loops upon failure.

# Tests

Verified autoregressive decoding execution, real-time subprocess logging, and JSON report generation across 4 distinct models. The script correctly enforces required overrides, parses custom GCS checkpoint/tokenizer paths, and uploads structured reports to GCS upon completion.

<details>
<summary><b>1. Qwen3-14b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/qwen3-14b/scanned/2026-04-12/0/items",
    "hf_config_url": "https://huggingface.co/Qwen/Qwen3-14B/raw/main/config.json",
    "hf_model_path": "Qwen/Qwen3-14B",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/qwen3/modeling_qwen3.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "qwen3-14b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": true,
        "tokenizer_path": "Qwen/Qwen3-14B",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-qwen3-14b-full-run"
}
```

**Decoding Validation Task Execution Log:**
https://paste.googleplex.com/5045773933871104
</details>  

<details>
<summary><b>2. Mistral-7b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/mistral-7b/2025-01-23-19-04/unscanned/checkpoints/0/items",
    "hf_config_url": "https://huggingface.co/mistralai/Mistral-7B-v0.1/raw/main/config.json",
    "hf_model_path": "mistralai/Mistral-7B-v0.1",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/mistral/modeling_mistral.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "mistral-7b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "dtype": "bfloat16",
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "mistralai/Mistral-7B-v0.1",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-mistral-7b-full-run"
}
```

**Decoding Validation Task Execution Log:**
https://paste.googleplex.com/4642963094372352
</details>  

<details>
<summary><b>3. Qwen3-8b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/qwen3-8b/unscanned/0/items",
    "hf_config_url": "https://huggingface.co/Qwen/Qwen3-8B/raw/main/config.json",
    "hf_model_path": "Qwen/Qwen3-8B",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/qwen3/modeling_qwen3.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "qwen3-8b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "Qwen/Qwen3-8B",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-qwen3-8b-full-run"
}
```

**Decoding Validation Task Execution Log:**
https://paste.googleplex.com/6236433022058496
</details>  

<details>
<summary><b>4. Gemma3-4b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/gemma3-4b/2025-03-18-19-03/unscanned/checkpoints/0/items",
    "hf_config_url": "https://huggingface.co/google/gemma-3-4b-it/raw/main/config.json",
    "hf_model_path": "google/gemma-3-4b-it",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/gemma3/modeling_gemma3.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "gemma3-4b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "google/gemma-3-4b-it",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-gemma3-4b-full-run"
}
```

**Decoding Validation Task Execution Log:**
https://paste.googleplex.com/4656868940185600
</details>  

# Stack
4th PR in stack. Depends on the preceding upstream PR: #4727 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by olufiyin19** [comment]
@gemini-cli /review
